core/src/main/java/io/opencensus/tags/TagKey.java

```diff
@@ -14,92 +14,189 @@
+   * <pre>{@code
+   * Tag tag =
+   *     tagKey.matchWithDefault(
+   *         stringKey -> TagString.create(stringKey, TagValueString.create("string value")),
+   *         longKey -> TagLong.create(longKey, 100L),
+   *         booleanKey -> TagBoolean.create(booleanKey, true),
+   *         unknownKey -> {
+   *           logger.log(Level.WARNING, "Unknown tag key type: " + unknownKey.toString());
+   *           return TagString.create(
+   *               TagKeyString.create(unknownKey.getName()),
+   *               TagValueString.create("string value"));
+   *         });
+   * }</pre>
    *
-   * <ol>
-   * <li>It cannot be longer than {@link #MAX_LENGTH}.
-   * <li>It can only contain printable ASCII characters.
-   * </ol>
-   *
-   * @param name the name of the key.
-   * @throws IllegalArgumentException if the name is not valid.
+   * @param stringFunction the function to call when the {@code TagKey} is a {@code TagKeyString}.
+   * @param longFunction the function to call when the {@code TagKey} is a {@code TagKeyLong}.
+   * @param booleanFunction the function to call when the {@code TagKey} is a {@code TagKeyBoolean}.
+   * @param defaultFunction the function to call when the tag key has a type other than {@code
+   *     String}, {@code long}, or {@code boolean}.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag key's type.
    */
-  // TODO(sebright): Make this public once we support types other than String.
-  static TagKey<Long> createLongKey(String name) {
-    checkArgument(StringUtil.isValid(name));
-    return new AutoValue_TagKey<Long>(name, TAG_LONG);
+  // TODO(sebright): Should we make this public in the first release?
+  // TODO(sebright): How should we deal with the possibility of adding more tag types in the future?
+  //                 Will this method work? What type of parameter should we use for the default
+  //                 case? Should we remove the non-backwards compatible "match" method?
+  public abstract <T> T matchWithDefault(
+      Function<? super TagKeyString, T> stringFunction,
+      Function<? super TagKeyLong, T> longFunction,
+      Function<? super TagKeyBoolean, T> booleanFunction,
+      Function<? super TagKey, T> defaultFunction);
+
+  /** A {@code TagKey} for values of type {@code String}. */
+  @Immutable
+  @AutoValue
+  public abstract static class TagKeyString extends TagKey {
+
+    /**
+     * Constructs a {@code TagKeyString} with the given name.
+     *
+     * <p>The name must meet the following requirements:
+     *
+     * <ol>
+     *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+     *   <li>It can only contain printable ASCII characters.
+     * </ol>
+     *
+     * @param name the name of the key.
+     * @return a {@code TagKeyString} with the given name.
+     * @throws IllegalArgumentException if the name is not valid.
+     */
+    public static TagKeyString create(String name) {
```

Comment:
I think in this case we want to expose TagKeyString to the users to make them aware of the type.  Should we do then the same in Tag (then return a TagString)?